### PR TITLE
SERVERLESS-2327 Map OW env to Lambda context for Go

### PR DESCRIPTION
**Pull request by mattwelke**
Adds the context mapping we need for our plans to support Lambda-like functions to our Go runtime. In order to add that functionality to all Go runtimes we support in production today, that meant also backporting what we we were already doing for Go 1.17 to 1.15 too. Until now, the 1.15 runtime just supports the map/map signature. So, that's done here too.

The tests for the context mapping feature and for verifying that all signatures are supported in both the Go 1.15 and Go 1.17 runtimes are in the serverless/main repo.

### CODE REVIEW

**Comment by mattwelke** [comment]
I could use some help adding tests for this.
I looked at launcher_test.go but concluded that because nothing there involved setting up a container, that it wouldn't work for what I want to test here since I want to test how env vars are set to ensure that the context is set up correctly.

I looked at the Scala test file but I don't know how to wire things up. I added a test that you can see in this WIP PR but because I'm not overriding anything, I don't think it will end up executing.

Edit: Ended up adding tests only in Go to the serverless/main repo.

File: golang1.17/lib/launcher.go [hunk 1]
```diff
@@ -206,6 +206,9 @@ func buildArguments(ctx context.Context, f reflect.Value, in []byte) ([]reflect.
 		return nil, nil
 	}
 
+	// Only build the context if the customer provided enough parameters in their function to require one.
+	ctx = buildContext(ctx)
```
Comment (markusthoemmes) [inline golang1.17/lib/launcher.go]: Why not call this above, where the deadline is put into the context as well?

File: golang1.17/lib/launcher.go [hunk 2]
```diff
@@ -227,6 +230,37 @@ func buildArguments(ctx context.Context, f reflect.Value, in []byte) ([]reflect.
...
+// so that you can differentiate constant instances of the structs.
+// See https://github.com/golang/go/blob/master/src/net/http/server.go for an example of this.
+//
+// At DO, we don't yet offer a library our users can import as they write their functions, so we can't follow this
+// pattern right now. Therefore, the naming convention we use for our context keys doesn't matter. In the future, we can
+// change this by creating a library that our customers can use, at which point we will no longer have multiple context
+// keys and the single remaining context key will no longer be a string. Therefore, while we do need to communicate
+// these string keys to our customers for now, the naming convention we use for the strings doesn't matter. We've chosen
+// this-case arbitrarily.
+func buildContext(ctx context.Context) context.Context {
```
Comment (markusthoemmes) [inline golang1.17/lib/launcher.go]: Should we subsume adding the deadline to the context into this function as well?

File: golang1.17/lib/launcher.go (hunk 1 above)
Comment (mattwelke) [reply golang1.17/lib/launcher.go]: Decided to refactor. See below.

File: golang1.17/lib/launcher.go (hunk 2 above)
Comment (mattwelke) [reply golang1.17/lib/launcher.go]: Agree. I moved all of the context building into one place. To keep things as idiomatic as possible for Go, I have the context built in the `execute` function. I feel like it's idiomatic to construct a context, with a deadline if you need it, before you use the context in a function call you might need to cancel.

**Comment by mattwelke** [comment]
Ready for review again. The tests are in `language_tests.go` in serverless/main. There's some repeated coded here (in order to backport the context mapping to Go 1.15), but I think that's fine for now. We have plans to consolidate all code into serverless/main, and maybe we could address code re-use at that point.

File: golang1.15/lib/launcher.go [hunk 3]
```diff
@@ -20,17 +20,26 @@ package main
 import (
```
Comment (markusthoemmes) [inline golang1.15/lib/launcher.go]: We don't need to backport to 1.15. We won't document it.

File: golang1.17/lib/launcher.go [hunk 4]
```diff
@@ -118,15 +118,11 @@ func execute(f interface{}, in []byte) ([]byte, error) {
 		}
```
Comment (markusthoemmes) [inline golang1.17/lib/launcher.go:118]: This piece LGTM.

**Review by markusthoemmes — CHANGES_REQUESTED** [review]

File: golang1.17/lib/launcher.go [hunk 5]
```diff
@@ -227,6 +223,52 @@ func buildArguments(ctx context.Context, f reflect.Value, in []byte) ([]reflect.
...
+	ctx := context.Background()
+
+	var cancel context.CancelFunc
+	if deadline := os.Getenv("__OW_DEADLINE"); deadline != "" {
+		// Set up a context that cancels at the given deadline.
+		deadlineMillis, err := strconv.ParseInt(os.Getenv("__OW_DEADLINE"), 10, 64)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to parse deadline: %w", err)
+		}
+		ctx, cancel = context.WithDeadline(context.Background(), time.UnixMilli(deadlineMillis))
```
Comment (markusthoemmes) [inline golang1.17/lib/launcher.go]:
```suggestion
		ctx, cancel = context.WithDeadline(ctx, time.UnixMilli(deadlineMillis))
```

File: golang1.15/lib/launcher.go (hunk 3 above)
Comment (mattwelke) [reply golang1.15/lib/launcher.go]: Reverted 1.15 code back.

**Review by markusthoemmes — APPROVED** [review]

**Pull request merged**